File: java/app/src/main/java/studentdb/model/CourseAttendanceData.java
package studentdb.model;

/**
 * Represents the status of a student's enrollment in a course, including attendance status and
 * optional grade.
 *
 * @param status the attendance status (required)
 * @param grade the grade earned (required only when status is FINISHED, otherwise must be null)
 */
public record CourseAttendanceData(StatusData status, CourseAttendanceDetailsGrade grade) {

  /** Represents a grade earned in a completed course. */
  public enum CourseAttendanceDetailsGrade {
    A,
    B,
    C,
    D,
    F
  }

  public CourseAttendanceData {
    if (status == null) {
      throw new IllegalArgumentException("Attendance status cannot be null");
    }

    if (status == StatusData.FINISHED && grade == null) {
      throw new IllegalArgumentException("Grade is required when status is FINISHED");
    }

    if (status != StatusData.FINISHED && grade != null) {
      throw new IllegalArgumentException(
          "Grade must be null when status is not FINISHED (status: " + status + ")");
    }
  }
}

File: python/studentdb/model/CourseAttendanceData.py
from dataclasses import dataclass
from enum import Enum
from typing import Optional

from studentdb.model.StatusData import StatusData


class CourseAttendanceDetailsGrade(Enum):
    """Represents a grade earned in a completed course."""
    A = "A"
    B = "B"
    C = "C"
    D = "D"
    F = "F"


@dataclass(frozen=True)
class CourseAttendanceData:
    """
    Represents the status of a student's enrollment in a course, including attendance status and
    optional grade.

    Args:
        status: the attendance status (required)
        grade: the grade earned (required only when status is FINISHED, otherwise must be None)
    """
    status: StatusData
    grade: Optional[CourseAttendanceDetailsGrade]

    def __post_init__(self):
        if self.status is None:
            raise ValueError("Attendance status cannot be None")

        if self.status == StatusData.FINISHED and self.grade is None:
            raise ValueError("Grade is required when status is FINISHED")

        if self.status != StatusData.FINISHED and self.grade is not None:
            raise ValueError(
                f"Grade must be None when status is not FINISHED (status: {self.status})"
            )
